lib/shoryuken/processor.rb

```diff
@@ -1,15 +1,79 @@
-require 'json'
-
 module Shoryuken
   class Processor
-    include Celluloid
     include Util
 
-    def initialize(manager)
-      @manager = manager
+    BATCH_LIMIT = 10
+
+    def initialize(fetcher, polling_strategy)
+      @fetcher = fetcher
+      @polling_strategy = polling_strategy
+      # what is the worst thing that could happen because it's not atomic?
+      # is it possible that it evaluates to true even if we never set it to true?
+      @done = false
+    end
+
+    def start
+      # does updating and evaluating @thread has to be wrapped in mutex?
+      # can it throw exception?
+      # what should we do if it dies?
+      @thread ||= spawn_thread(&method(:loop))
+    end
+
+    def terminate(wait=false)
+      @done = true
+      # does updating and evaluating @thread has to be wrapped in mutex?
+      return if !@thread
+      @thread.value if wait
+    end
+
+    def kill
+      @done = true
+      # does updating and evaluating @thread has to be wrapped in mutex?
+      return if !@thread
+      @thread.raise ::Shoryuken::Shutdown
+    end
```

Comment:
@mariokostelac I'm not against at all, I'm just reviewing it. Sorry, it may look like I'm just being negative, but that's not the idea.

> That way manager becomes simple, we do not share state - no concurrency problems, and we do not have to handle threads ourselves.

That's the trade off we need to think about. If we are going with this approach, we can have self-sufficient processors, but they will need to fetch a single message, process, then fetch a single one again. We will increase the "parallelism", but increase the number the requests a lot, which consumes time and $. We need to balance out, if trying to fetch as much as we can per second, is better or not. TBH, I don't think it will be easy to find an exact answer.